ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/commands/tracker/CommandTracker.scala

```diff
@@ -275,39 +293,91 @@ private[commands] class CommandTracker[Context](
               Nil
             }
           }
-          .to(immutable.Seq)
+          .to(immutable.Iterable)
       }
 
-      private def getOutputForCompletion(completion: Completion) = {
-        val (commandId, errorText) = {
-          completion.status match {
-            case Some(StatusProto(code, _, _, _)) if code == Status.Code.OK.value =>
-              completion.commandId -> "successful completion of command"
-            case _ =>
-              completion.commandId -> "failed completion of command"
-          }
+      private def getOutputForCompletion(
+          completion: Completion
+      ): immutable.Iterable[Ctx[Context, Either[CompletionFailure, CompletionSuccess]]] = {
+        val completionDescription = completion.status match {
+          case Some(StatusProto(code, _, _, _)) if code == Status.Code.OK.value =>
+            "successful completion of command"
+          case _ => "failed completion of command"
         }
 
-        logger.trace("Handling {} {}", errorText, completion.commandId: Any)
-        pendingCommands.remove(commandId).map { t =>
-          Ctx(t.context, tracker.CompletionResponse(completion))
+        val commandId = completion.commandId
+        logger.trace(
+          "Handling {} {} from submission {}.",
+          completionDescription,
+          commandId,
+          completion.submissionId,
+        )
+
+        val maybeSubmissionId = Option(completion.submissionId).collect {
+          case id if id.nonEmpty => id
+        }
+        val trackedCommands =
+          pendingCommandKeys(maybeSubmissionId, commandId).flatMap(pendingCommands.remove(_).toList)
+
+        if (trackedCommands.size > 1) {
+          trackedCommands.map { trackingData =>
+            Ctx(
+              trackingData.context,
+              Left(
+                NotOkResponse(
+                  commandId,
+                  StatusProto.of(
+                    Status.Code.INTERNAL.value(),
+                    s"There are multiple pending commands for the id $commandId. This can only happen for the mutating schema.",
```

Comment:
```suggestion
                    s"There are multiple pending commands with ID $commandId for submission ID $maybeSubmissionId. This can only happen for the mutating schema that shouldn't be used anymore as it doesn't fully support command deduplication.",
```

Reply:
Or something more explanatory anyway.